docs/src/references/synchronize-daemons.md

```diff
@@ -0,0 +1,116 @@
+# Introduction
+This document explains the scenarios that require the startup of different tedge daemons to be synchronized,
+and solutions to avoid this synchronization by relying on dynamic runtime health status.
+
+# Problem statement
+
+Thin-edge.io has three main issues for synchronizing the daemons
+
+1. The thin-edge daemons depend on another daemon's liveness to delegate or request some work.
+ For example, tedge-mapper-c8y depends on tedge-agent. When the tedge-mapper-c8y is started,
+ the mapper sends a request to get the existing software list on `tedge/commands/req/software/list` topic,
+ if the tedge-agent is not up and running then the request message will be lost. So, it's better to check if the agent
+ is up and running before sending the request to get the software list.
+
+2. The thin-edge daemons are dependent on the liveness of the bridge to communicate with the cloud.
+ For example, when the tedge-mapper-c8y comes up, it will try to get the jwt token. The request will be sent to c8y cloud
+ over c8y/s/dat, if the bridge is not up then the request message will be lost.
+ So, better to check the status of the bridge before sending the request to the c8y cloud.
```

Comment:
> > when the tedge-mapper-c8y comes up, it will try to get the jwt token
> 
> Yes, but this is not the main point: what the c8y mapper needs is the c8y internal id of the device and this internal id is itself only required to use the c8y REST API (say to send a large event which payload doesn't fit in an MQTT message).

Agreed, but the `Internal ID` is retrieved only once and this is re-used throughout the mapper life cycle right?.

> 
> > So, better to check the status of the bridge before sending the request to the c8y cloud.
> 
> Yes, definitely. For instance, sending the list of software must be done not when the mapper start but each time the bridge to c8y is established. A bit more complex, because the agent must also be up.

The liveness of the `c8y` bridge and `tedge-agent` is not checked at the same time, before forwarding the request to the tedge-agent the liveness will be checked, and the `c8y` bridge liveness is required only when the mapper has to communicate with the cloud.
 
> Unfortunately, things are a bit more complex when the bridge is down: some capabilities are no more available. This is not an issue when the requests are triggered from the cloud (as for the operations). It's an issue when the mapper uses HTTP instead of MQTT (e.g for large events - because the events sent over HTTP will be lost when the bridge is down and not persisted as for those sent over MQTT). This issue is far beyond the scope of this proposal.

The whole proposal is only for the MQTT base communication.

Reply:
> Agreed, but the Internal ID is retrieved only once and this is re-used throughout the mapper life cycle right?.

Yes. However, getting this internal id on start is where all the synchronization issues have been observed.